astro/src/content/blog/forgerock-password-stroge.mdx

```diff
@@ -0,0 +1,114 @@
+
+## How Do I Decipher Forgerock's Hash
+
+Password hash and salt storage will vary between implementations. Hashing algorithms will vary as well and could even vary between users. By default, Forgerock uses the PBKDF2-HMAC-SHA256 encryption with their Cloud Development Kit so that will be used in this article.
+
+<Aside type="note">
+If your implementation is using a different hashing function, Forgerock has more details on their other password storage schemes [here](https://backstage.forgerock.com/knowledge/kb/article/a44757687).
+</Aside>
+
+Forgerock stores that hash in the following way: 
+
+"&#123;PBKDF2-HMAC-SHA256&#125;"&lt;iterations&gt;":" base64(&lt;digest&gt; &lt;salt&gt;) where:
+
+- &#123;PBKDF2-HMAC-SHA256&#125; is a literal describing the hashing function used.
+- &lt;iterations&gt; is the number of times the algorithm is applied
+- ":" is a literal separator
+- base64(&lt;digest&gt; &lt;salt&gt;) indicates that the password hash and salt are stored together and base64 encoded
+
+In the example:
+
+&#123;PBKDF2-HMAC-SHA256&#125;10:8c7nLGEIXeZf45YQ92A2MD+v8olvKKl6iWXGQZoluJ/awqZnHwFvslIOx7xOZ9AV
+  
+  - PBKDF2-HMAC-SHA256 is the hashing function used
+  - 10 is the number of iterations
+  - 8c7nLGEIXeZf45YQ92A2MD+v8olvKKl6iWXGQZoluJ/awqZnHwFvslIOx7xOZ9AV is the Base64 encoded password hash and salt
+
+So how do you decipher the password hash from the salt? The first thing you need to do is Base64 decode the string. This give you:
+
+ ñ�￹ç,a￹]æ_ã�￹÷`60?¯ò�￹o(©z�￹e�￹A�￹%¸�￹�￹�￹¦g￹o²R￹�￹¼Ng�
+
+ ![What the heck is that?](/img/blogs/forgerock-password-storage/what-the-heck.jpg)
+ 
+ I know, not very human readable. From here, you need to write some code to get the values you are looking for. Below is an excerpt from the script used in FusionAuth's [Migration From Forgerock](/docs/lifecycle/migrate-users/bulk/forgerock) guide.
+
+<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-import-scripts/master/forgerock/import.rb" lang="ruby" tags="ForgerockPasswordHash" />
+
+This is Ruby code, but you can implement it in the language of your choice. 
+
+Breaking the code down, you can see that it starts by accepting the 'hashstring' as an argument. In this case, the 'hashstring' will look like our example from above, &#123;PBKDF2-HMAC-SHA256&#125;10:8c7nLGEIXeZf45YQ92A2MD+v8olvKKl6iWXGQZoluJ/awqZnHwFvslIOx7xOZ9AV. 
+
+Next, it uses regular expressions to extract each part of the stored password information. At this point you should have the encryption scheme (PBKDF2-HMAC-SHA256) and number of iterations (10) in plain text. 
+
+It then takes the base64(&lt;digest&gt; &lt;salt&gt;) part of the stored information, `password_salt_hash`, and decodes it from Base64. Once the decoding is done, it will then take the appropriate number of bytes of the decoded string and put it into variables, `hash_bytes` and `salt_bytes`. From there, it re-encodes those bytes back to Base64.
```

Comment:
```suggestion
It then takes the `base64(&lt;digest&gt; &lt;salt&gt;)` part of the stored information, which is in the `password_salt_hash` variable, and decodes it from Base64. Once the decoding is done, it will then take the appropriate number of bytes of the decoded string and put it into variables, `hash_bytes` and `salt_bytes`. 

From there, it re-encodes those bytes back to Base64, which is what FusionAuth needs for importing. Consult documentation of your destination identity store for the correct format for these values.
```

Reply:
"the appropriate number of bytes" --- where did you get this from? From ForgeRock's documentation?

Reply:
Made more clear.